### PR TITLE
Add DeviceInfo to NetworkStatus

### PR DESCRIPTION
**Pull request by Billy99**
Update the NetworkStatus with DeviceInfo per the approved device-info-spec, which can be found:
https://github.com/k8snetworkplumbingwg/device-info-spec

This allows DevicePlugins and CNIs to add additional data to the NetworkStatus annotation, which can be used by a container to add in deployment of the container workload.

### CODE REVIEW

**Comment by Billy99** [comment]
@amorenoz PTAL

File: pkg/apis/k8s.cni.cncf.io/v1/types.go [hunk 1]
```diff
@@ -38,15 +38,52 @@ type DNS struct {
...
+}
+
+type PciDevice struct {
+	PciAddress   string `json:"pci-address,omitempty"`
+	Vhostnet     string `json:"vhost-net,omitempty"`
+	RdmaDevice   string `json:"rdma-device,omitempty"`
+	PfPciAddress string `json:"pf-pci-address,omitempty"`
+}
+
+type VdpaDevice struct {
```
[the remark this answers is not in the document]
Comment (Billy99) [reply pkg/apis/k8s.cni.cncf.io/v1/types.go:59]: Done.

File: pkg/utils/cniconfig.go [hunk 2]
```diff
@@ -117,3 +126,112 @@ func GetCNIConfigFromSpec(configData, netName string) ([]byte, error) {
...
+func getDPDeviceInfoPath(resourceName string, deviceID string) string {
+	return filepath.Join(baseDevInfoPath, dpDevInfoSubDir, fmt.Sprintf("%s-%s-device.json",
+		strings.ReplaceAll(resourceName, "/", "-"), strings.ReplaceAll(deviceID, "/", "-")))
+}
+
+// getCNIDeviceInfoPath returns the standard Device Plugin DevInfo filename
+// The path is fixed but the filename is flexible and determined by the caller.
+func getCNIDeviceInfoPath(filename string) string {
+	return filepath.Join(baseDevInfoPath, cniDevInfoSubDir, filename)
+}
```
[the remark this answers is not in the document]
Comment (Billy99) [reply pkg/utils/cniconfig.go:196]: Done.

File: pkg/utils/cniconfig.go [hunk 3]
```diff
@@ -117,3 +126,112 @@ func GetCNIConfigFromSpec(configData, netName string) ([]byte, error) {
...
+	if err != nil {
+		return err
+	}
+	return saveDeviceInfo(devInfo, getCNIDeviceInfoPath(cniFilename))
+}
+
+// CleanDeviceInfoForCNI removes a DeviceInfo CNI File.
+func CleanDeviceInfoForCNI(cniFilename string) error {
+	return cleanDeviceInfo(getCNIDeviceInfoPath(cniFilename))
+}
```
[the remark this answers is not in the document]
Comment (Billy99) [reply pkg/utils/cniconfig.go:237]: Done. Changed all CNI functions to take cniPath.

**Review by Billy99 — COMMENTED** [review]
All comments addressed.

**Comment by Billy99** [comment]
@dougbtv @dcbw  PTAL

File: pkg/utils/net-attach-def.go [hunk 4]
```diff
@@ -122,7 +122,7 @@ func GetNetworkStatus(pod *corev1.Pod) ([]v1.NetworkStatus, error) {
 }
 
 // CreateNetworkStatus create NetworkStatus from CNI result
-func CreateNetworkStatus(r cnitypes.Result, networkName string, defaultNetwork bool) (*v1.NetworkStatus, error) {
+func CreateNetworkStatus(r cnitypes.Result, networkName string, defaultNetwork bool, dev *v1.DeviceInfo) (*v1.NetworkStatus, error) {
```
Comment (dougbtv) [inline pkg/utils/net-attach-def.go:125]: This is a totally valid change, and I'm 100% OK with it. 

But it makes me realize that we don't yet have any tags for this repo. So, with a change for a publicly available method... I think we might want to tag a version before we merge this in. I'm adding some notes to the NPWG agenda to see where we should start with tagging a version.

File: pkg/apis/k8s.cni.cncf.io/v1/types.go [hunk 5]
```diff
@@ -38,15 +38,53 @@ type DNS struct {
...
-	Mac       string   `json:"mac,omitempty"`
-	Default   bool     `json:"default,omitempty"`
-	DNS       DNS      `json:"dns,omitempty"`
+	Name       string     `json:"name"`
+	Interface  string     `json:"interface,omitempty"`
+	IPs        []string   `json:"ips,omitempty"`
+	Mac        string     `json:"mac,omitempty"`
+	Default    bool       `json:"default,omitempty"`
+	DNS        DNS        `json:"dns,omitempty"`
+	DeviceInfo DeviceInfo `json:"device-info,omitempty"`
```
Comment (dcbw) [inline pkg/apis/k8s.cni.cncf.io/v1/types.go:87]: Should DeviceInfo be a pointer? that way we can check for `nil` and that might be cleaner than "if status.DeviceInfo.Type == xxx`

Comment (Billy99) [reply pkg/apis/k8s.cni.cncf.io/v1/types.go:87]: Done.

File: pkg/utils/net-attach-def.go (hunk 4 above)
Comment (dougbtv) [reply pkg/utils/net-attach-def.go:125]: Added release v1.1.0 @ https://github.com/k8snetworkplumbingwg/network-attachment-definition-client/releases/tag/v1.1.0

**Review by dougbtv — APPROVED** [review]
LGTM!

**Pull request merged**